src/front/glsl/preprocess.rs

```diff
@@ -0,0 +1,145 @@
+use crate::FastHashMap;
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum Error {
+    #[error("unmatched else")]
+    UnmatchedElse,
+    #[error("unmatched endif")]
+    UnmatchedEndif,
+}
+
+#[derive(Clone, Debug)]
+pub struct LinePreProcessor {
+    pub defines: FastHashMap<String, String>,
+    if_stack: Vec<(bool, bool)>,
```

Comment:
this looks quite weird, it's totally unclear what this `(bool, bool)` is

Reply:
Indeed, will try turning into small struct instead.